File: psana_examples/tags/V00-00-09/src/DumpAcqiris.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpAcqiris...
//
// Author List:
//      Andrei Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpAcqiris.h"

//-----------------
// C/C++ Headers --
//-----------------

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "pdsdata/xtc/DetInfo.hh"
#include "psddl_psana/acqiris.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpAcqiris)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpAcqiris::DumpAcqiris (const std::string& name)
  : Module(name)
{
  // get the values from configuration or use defaults
  m_src = configStr("source", "DetInfo(:Acqiris)");
}

//--------------
// Destructor --
//--------------
DumpAcqiris::~DumpAcqiris ()
{
}

/// Method which is called at the beginning of the calibration cycle
void 
DumpAcqiris::beginCalibCycle(Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  shared_ptr<Psana::Acqiris::ConfigV1> acqConfig = env.configStore().get(m_src);
  if (acqConfig.get()) {
    MsgLog(name(), info, "Acqiris::ConfigV1: nbrBanks=" << acqConfig->nbrBanks()
           << " channelMask=" << acqConfig->channelMask()
           << " nbrChannels=" << acqConfig->nbrChannels()
           << " h.sampInterval=" << acqConfig->horiz().sampInterval()
           << " h.delayTime=" << acqConfig->horiz().delayTime()
           << " h.nbrSamples=" << acqConfig->horiz().nbrSamples()
           << " h.nbrSegments=" << acqConfig->horiz().nbrSegments()
           << " v[0].fullScale=" << acqConfig->vert(0).fullScale()
           << " v[0].offset=" << acqConfig->vert(0).offset()
           << " v[1].fullScale=" << acqConfig->vert(1).fullScale()
           << " v[1].offset=" << acqConfig->vert(1).offset()
           );
  }
}

/// Method which is called with event data, this is the only required 
/// method, all other methods are optional
void 
DumpAcqiris::event(Event& evt, Env& env)
{

  shared_ptr<Psana::Acqiris::DataDescV1> acqData = evt.get(m_src);
  if (acqData.get()) {
    const std::vector<int>& shape = acqData->data_shape();
    for (int i = 0; i < shape[0]; ++ i) {
      const Psana::Acqiris::DataDescV1Elem& elem = acqData->data(i);
      MsgLog(name(), info, "Acqiris::DataDescV1: element=" << i 
           << " nbrSegments=" << elem.nbrSegments()
           << " nbrSamplesInSeg= " << elem.nbrSamplesInSeg());
    }
  }
}

} // namespace psana_examples

File: python_examples/trunk/src/dump_acqiris.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id: dump_acqiris.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $
#
# Description:
#  Pyana user analysis module dump_princeton...
#
#------------------------------------------------------------------------

"""Example module for accessing SharedIpimb data.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@version $Id: dump_acqiris.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $

@author Andy Salnikov
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision: 2622 $"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import xtc

#----------------------------------
# Local non-exported definitions --
#----------------------------------

# local definitions usually start with _

#---------------------
#  Class definition --
#---------------------
class dump_acqiris (object) :
    """Class whose instance will be used as a user analysis module. """

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, source="" ) :
        self.m_src = source

    #-------------------
    #  Public methods --
    #-------------------
    def beginjob( self, evt, env ) :
        
        src = env.configStr("source", "DetInfo(:Acqiris)")
        print 'env.configStr("source", "DetInfo(:Acqiris)") ->', src

        foundSrc_list = []
        config = env.configStore().get("Psana::Acqiris::ConfigV1", env.Source(src), foundSrc_list);
        foundSrc = foundSrc_list[0]
        print "foundSrc: log()=%x, phy()=%x" % (foundSrc.log(), foundSrc.phy())
        if config:
        
            print "%s: %s" % (config.__class__.__name__, self.m_src)
            
            print "  nbrBanks =", config.nbrBanks(),
            print "channelMask =", config.channelMask(),
            print "nbrChannels =", config.nbrChannels(),
            print "nbrConvertersPerChannel =", config.nbrConvertersPerChannel()
     
            h = config.horiz()
            print "  horiz: sampInterval =", h.sampInterval(),
            print "delayTime =", h.delayTime(),
            print "nbrSegments =", h.nbrSegments(),
            print "nbrSamples =", h.nbrSamples()

            nch = config.nbrChannels()
            for ch in range(nch):
                v = config.vert()[ch]
                print "  vert(%d):" % ch,
                print "fullScale =", v.fullScale(),
                print "slope =", v.slope(),
                print "offset =", v.offset(),
                print "coupling =", v.coupling(),
                print "bandwidth=", v.bandwidth()


    def event( self, evt, env ) :
        """This method is called for every L1Accept transition.

        @param evt    event data object
        @param env    environment object
        """

        print "... self.m_src=", self.m_src
        acqData = evt.get("Psana::Acqiris::DataDesc", env.Source(self.m_src))

        nchan = acqData.data_shape()[0];
        for chan in range(nchan):
            elem = acqData.data(chan)

            print "%s: %s: channel = %d" % (elem.__class__.__name__, self.m_src, chan)
            print "  nbrSegments =", elem.nbrSegments()
            print "  nbrSamplesInSeg =", elem.nbrSamplesInSeg()
            print "  timestamps =", [elem.timestamp()[seg].pos() for seg in range(elem.nbrSegments())]
            wf = elem.waveforms()
            print "  waveform [len=%d] = %s" % (len(wf), wf)


    def endjob( self, evt, env ) :
        
        pass

